Add staking and operators guide (#199)
subspace/subspace-docs: docs/operators_and_staking/intro.md

- **Higher Hardware Requirements**: Operators require more substantial hardware capabilities, as they must execute complex transactions.
- **Pre-Validation and Batching**: Operators pre-validate and batch transactions into bundles through a stake-weighted election process.
- **Deterministic Execution**: The operators execute transactions in a specific, deterministic order, producing state commitments in the form of execution receipts.
- **Secondary Network Role**: Participates in a more deterministic process involving the execution of transactions, potentially requiring specialized infrastructure or capabilities.
- **Supports Various Environments**: Can support different smart contract execution environments like the Ethereum Virtual Machine (EVM) or Web-Assembly (WASM).

### Staking

The Subspace Network relies on staking from both domain operators and farmers to secure the network and provide resources. Subspace implements a Nominated Proof-of-Stake algorithm where farmers endorse operators who execute transactions and produce blocks.

Our staking model consists of two tiers:

> - Operators stake to gain the right to execute transactions within a domain and produce blocks. They are responsible for validating and executing transactions, producing execution receipts, and applying state transitions and earn rewards for their work. The operator chances to be elected as a slot leader are weighted by stake. Operators can be nominated by farmers.

Reviewer: Consider mentioning operator stake can be slashed which provides crypto-economic security to execution.

Reviewer: I don't believe [farmers are ever subject to slashing](https://github.com/subspace/subspace-docs/pull/199/commits/99cf0a961264c4830d9f60fbdfa0c96114854d07#diff-71f055037303c66063fa39a61f825280ed4be6b204aab5ec6c8a46f79a65c21eR43). It only applied to malicious operators. @jfrank-summit could you confirm?

Author: It may be the phrasing that's confusing. 

According to [Subnomicon](https://subnomicon.subspace.network/docs/decex/staking/) farmer's rewards can be automatically staked by the trusted operator. If the operator acts maliciously, staked tokens can be slashed, so the farmer can be slashed of tokens in case they nominate a suspicious or unfair operator.

Reviewer: I'm not sure this is quite right @EmilFattakhov. If I understand, the wording means that a farmer can send their rewards directly to an operator's stake from their farm but I don't think those tokens are ever at risk.

Reviewer 2: Sorry @jim-counter I didn't see your question before. Nominated tokens are in fact at risk of being slashed.

Reviewer: Thanks for confirming. I was not aware until now!

# Key differences between farming and being an operator

### Farming
- **Transaction Ordering**: Farmers are responsible for confirming the availability of transactions and providing an ordering.
- **Lightweight Requirements**: The hardware requirements for farming are designed to be lightweight, making it accessible to anyone.
- **Verification**: Farmers only verify the proof-of-election and ensure that the data is available.
- **Transactions**: Farmers do not execute transactions; they focus on ordering them and including them in the blockchain.

> - **Primary Role**: Focuses on the probabilistic process of achieving consensus over transaction ordering.

Author: Good point Justin and nicely phrased, love it!

### Being an operator

- **Transaction Execution**: Operators are responsible for executing transactions and maintaining the resulting chain state.
- **Higher Hardware Requirements**: Operators require more substantial hardware capabilities, as they must execute complex transactions.
- **Pre-Validation and Batching**: Operators pre-validate and batch transactions into bundles through a stake-weighted election process.
- **Deterministic Execution**: The operators execute transactions in a specific, deterministic order, producing state commitments in the form of execution receipts.
- **Secondary Network Role**: Participates in a more deterministic process involving the execution of transactions, potentially requiring specialized infrastructure or capabilities.
- **Supports Various Environments**: Can support different smart contract execution environments like the Ethereum Virtual Machine (EVM) or Web-Assembly (WASM).

> 

Reviewer: I'd suggest describing the slashable offence as being a malicious actor rather than behaving wrongly. The operator risks his stake to slashing if he acts maliciously. There is a big penalty to doing so. This provides crypto-economic security to execution.

> 

Author: Agree, that sounds like a valuable point to add to the list. Great feedback guys!

### Staking

The Subspace Network relies on staking from both domain operators and farmers to secure the network and provide resources. Subspace implements a Nominated Proof-of-Stake algorithm where farmers endorse operators who execute transactions and produce blocks.

Our staking model consists of two tiers:

- Operators stake to gain the right to execute transactions within a domain and produce blocks. They are responsible for validating and executing transactions, producing execution receipts, and applying state transitions and earn rewards for their work. The operator chances to be elected as a slot leader are weighted by stake. Operators can be nominated by farmers. 

- Farmers earn rewards proportional to their pledged storage. Farmers can choose to nominate operators and back them with their own stake, increasing their chance of being elected as a slot leader. Farmers, who have earned storage rewards, nominate operators to execute transactions. This nomination system balances the power between farmers and operators and both parties share the rewards and the potential penalties (slashing). 

### Stake Epoch

> StakeEpoch is a designated period in domain blocks within a blockchain system that marks each stake allocation re-adjustment period. Occurring every `StakeEpochDuration` blocks (at the moment, it's set to every 600 blocks or ~10 minutes), an epoch transition triggers specific actions such as finalizing operator domain switches, deregistering operators, unlocking operators and their associated funds, and recalculating stake distribution for the Verifiable Random Function (VRF) election. These transitions are designed to adjust the stake distribution dynamically, finalize various staking-related operations, process rewards, and manage deposits and withdrawals. The uniform duration across all domains helps maintain consistency in the network, while the specific starting point for each domain's epoch transition may vary based on when it is registered, helping to amortize the load of these transitions.

Reviewer: StakeEpoch -> Stake epoch

# Key differences between farming and being an operator

### Farming
- **Consensus:** This is the primary role of Farmers, and provides security and consensus for the network. Our Farmers are what ensure we don't trust, but verify.
- **Transaction Ordering**: Farmers are responsible for confirming the availability of transactions and providing an ordering.
- **Lightweight Requirements**: The hardware requirements for farming are designed to be lightweight, making it accessible to anyone.
- **Verification**: Farmers only verify the proof-of-election and ensure that the data is available.
- **Transactions**: Farmers do not execute transactions; they focus on ordering them and including them in the blockchain.

### Being an operator

- **Transaction Submission and Execution**: Operators are responsible for batching transactions into bundles and submitting them to the consensus chain, executing transactions included in the consensus block and maintaining the resulting chain state.

> - **Higher [Hardware Requirements](https://wiki.polkadot.network/docs/maintain-guides-how-to-validate-polkadot#standard-hardware)**: Operators require more substantial hardware capabilities, as they must execute complex transactions.

Author: I put the question at the engineering channel, will update it once I hear back.

- **Require Initial Investment:** Operators are required to stake a certain amount of SSC. If an operator acts maliciously, their stake is at risk of being **slashed**. Engaging in such malicious behavior carries significant penalties, providing crypto-economic security to execution.
- **Pre-Validation and Batching**: Operators pre-validate and batch transactions into bundles through a stake-weighted election process.
- **Deterministic Execution**: The operators execute transactions in a specific, deterministic order, producing state commitments in the form of execution receipts.
- **Secondary Network Role**: Monitors the Domain chain for malicious activity and submits fraud proofs to consensus chain.
- **Supports Various Environments**: Can support different smart contract execution environments like the Ethereum Virtual Machine (EVM) or Web-Assembly (WASM).

### Staking

> The Subspace Network relies on staking from both domain operators and farmers to secure the network and provide resources. Subspace implements a Nominated Proof-of-Stake algorithm where farmers endorse operators who execute transactions and produce blocks.

Reviewer: Token holders might be better.

Our staking model consists of two tiers:

- Farmers earn rewards proportional to their pledged storage. Farmers can choose to nominate operators and back them with their own stake, increasing their chance of being elected as a slot leader. Farmers, who have earned storage rewards, nominate operators to execute transactions. This nomination system balances the power between farmers and operators and both parties share the rewards and the potential penalties (slashing). 

> - Operators stake to gain the right to execute transactions within a domain and produce blocks. They are responsible for validating and executing transactions, producing execution receipts, and applying state transitions and earn rewards for their work. The operator's chances to be elected as a slot leader are weighted by stake. Operators can be nominated by farmers. 

Reviewer 2: From the spec, `If it is, the operator is a slot leader for that slot and can produce a bundle. They should generate a ProofOfElection.` so both are fair. Maybe the following change would clarify:
```suggestion
- Operators stake to gain the right to produce bundles within a domain. They are responsible for validating and executing transactions, producing execution receipts, and applying state transitions and earn rewards for their work. The operator's chances to be elected as a slot leader and produce a bundle are weighted by their stake. Operators can be nominated by farmers or other SSC holders. 
```

# Key differences between farming and being an operator

### Farming
- **Consensus:** This is the primary role of Farmers, and provides security and consensus for the network. Our Farmers are what ensure we don't trust, but verify.
- **Transaction Ordering**: Farmers are responsible for confirming the availability of transactions and providing an ordering.
- **Lightweight Requirements**: The hardware requirements for farming are designed to be lightweight, making it accessible to anyone.
- **Verification**: Farmers only verify the proof-of-election and ensure that the data is available.
- **Transactions**: Farmers do not execute transactions; they focus on ordering them and including them in the blockchain.
- **Primary Role**: Focuses on the probabilistic process of achieving consensus over transaction ordering.

### Being an operator

- **Transaction Execution**: Operators are responsible for executing transactions and maintaining the resulting chain state.

> - **Higher Hardware Requirements**: Operators require more substantial hardware capabilities, as they must execute complex transactions.

Author: I think it's a good idea, linked Poladot's hardware requirements.

- **Higher [Hardware Requirements](#operator-hardware-requirements)**: Operators require more substantial hardware capabilities, as they must execute complex transactions.
- **Require Initial Investment:** Operators are required to stake a certain amount of SSC. If an operator acts maliciously, their stake is at risk of being **slashed**. Engaging in such malicious behavior carries significant penalties, providing crypto-economic security to execution.
- **Pre-Validation and Batching**: Operators pre-validate and batch transactions into bundles through a stake-weighted election process.
- **Deterministic Execution**: The operators execute transactions in a specific, deterministic order, producing state commitments in the form of execution receipts.
- **Secondary Network Role**: Monitors the Domain chain for malicious activity and submits fraud proofs to consensus chain.
- **Supports Various Environments**: Can support different smart contract execution environments like the Ethereum Virtual Machine (EVM) or Web-Assembly (WASM).

### Operator hardware requirements

:::note

> The hardware requirements have not been benchmarked, and these are our best estimates. We're open to your feedback if you feel that the requirements listed here are too high or too low. 

Reviewer: ```suggestion
The hardware requirements have not been benchmarked, and these are our best estimates. We would appreciate your feedback if you feel that the requirements listed here are too high or too low. 
```

Author: Thanks Jim, rephrased.
